caas/kubernetes/provider/k8s.go

```diff
@@ -85,9 +108,229 @@ func (k *kubernetesClient) EnsureOperator(appName, agentPath string, config *caa
 	return k.createPod(pod)
 }
 
+// DeleteService deletes the specified service.
+func (k *kubernetesClient) DeleteService(appName string) (err error) {
+	logger.Debugf("deleting application %s", appName)
+
+	if err := k.deleteService(appName); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(k.deleteDeployment(appName))
+}
+
+// EnsureService creates or updates a service for pods with the given spec.
+func (k *kubernetesClient) EnsureService(appName, unitspec string, numUnits int32, config caas.ResourceConfig) (err error) {
+	logger.Debugf("creating/updating application %s", appName)
+
+	if numUnits <= 0 {
+		return errors.Errorf("number of units must be > 0")
+	}
+	if unitspec == "" {
+		return errors.Errorf("missing container spec")
+	}
+
+	var cleanups []func()
+	defer func() {
+		if err == nil {
+			return
+		}
+		for _, f := range cleanups {
+			f()
+		}
+	}()
+
+	unitSpec, err := parseUnitSpec(unitspec)
+	if err != nil {
+		return errors.Annotatef(err, "parsing unit spec for %s", appName)
+	}
+	if err := k.configureDeployment(appName, unitSpec, &numUnits); err != nil {
+		return errors.Annotate(err, "creating or updating deployment controller")
+	}
+	cleanups = append(cleanups, func() { k.deleteDeployment(appName) })
+
+	var port *v1.ContainerPort
```

Comment:
as discussed on IRC, I think we should expose all ports in the service, and just have the first one be the one that's exposed via ingress

Reply:
done